wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.0.0-beta4.0/1__add_devfile.sql

```diff
@@ -0,0 +1,267 @@
+CREATE TABLE devfile_endpoint_attributes (
+    devfile_endpoint_id    BIGINT,
+    value                  TEXT,
+    name                   VARCHAR(255)
+);
+-- constraints & indexes
+ALTER TABLE devfile_endpoint_attributes ADD CONSTRAINT fk_devfile_endpoint_attributes_id FOREIGN KEY (devfile_endpoint_id) REFERENCES devfile_endpoint (id);
+CREATE UNIQUE INDEX index_devfile_endpoint_attributes_name ON devfile_endpoint_attributes (devfile_endpoint_id, name);
+CREATE INDEX index_devfile_endpoint_attributes ON devfile_endpoint_attributes (devfile_endpoint_id);
+
+---------
+
+-- devfile component env
+
+CREATE TABLE devfile_env (
+    id                    BIGINT       NOT NULL,
+    name                  VARCHAR(255) NOT NULL,
+    value                 TEXT,
+    devfile_component_id  BIGINT,
+
+    PRIMARY KEY (id)
+);
+-- constraints & indexes
+ALTER TABLE devfile_env ADD CONSTRAINT fk_devfile_env_id FOREIGN KEY (devfile_component_id) REFERENCES devfile_component (id);
+CREATE UNIQUE INDEX index_devfile_env_component_name ON devfile_env (devfile_component_id, name);
+CREATE INDEX index_component_env_id ON devfile_env (devfile_component_id);
+
+---------
+
+-- devfile component volume
+
+CREATE TABLE devfile_volume (
+    id                     BIGINT       NOT NULL,
+    name                   VARCHAR(255) NOT NULL,
+    container_path         TEXT,
+    devfile_component_id   BIGINT,
+
+    PRIMARY KEY (id)
+);
+-- constraints & indexes
+ALTER TABLE devfile_volume ADD CONSTRAINT fk_devfile_volumes_id FOREIGN KEY (devfile_component_id) REFERENCES devfile_component (id);
+CREATE UNIQUE INDEX index_devfile_volume_component_name ON devfile_volume (devfile_component_id, name);
+CREATE INDEX index_component_volume_id ON devfile_volume (devfile_component_id);
+
+---------
+
+-- devfile component entrypoint
+
+CREATE TABLE devfile_entrypoint (
+    id                         BIGINT       NOT NULL,
+    parent_name                VARCHAR(255) NOT NULL,
+    container_name             VARCHAR(255) NOT NULL,
+    devfile_component_id       BIGINT,
+
+    PRIMARY KEY (id)
+);
+-- constraints & indexes
+ALTER TABLE devfile_entrypoint ADD CONSTRAINT fk_devfile_entrypoints_id FOREIGN KEY (devfile_component_id) REFERENCES devfile_component (id);
+CREATE UNIQUE INDEX index_devfile_entrypoint_id_parent_container ON devfile_entrypoint (devfile_component_id, parent_name, container_name);
+CREATE INDEX index_component_entrypoint_id ON devfile_entrypoint (devfile_component_id);
+
+
+CREATE TABLE entrypoint_arg (
+    entrypoint_id    BIGINT,
+    args             VARCHAR(255) NOT NULL
+);
+
+-- constraints & indexes
+ALTER TABLE entrypoint_arg ADD CONSTRAINT fk_entrypoint_arg_id FOREIGN KEY (entrypoint_id) REFERENCES devfile_entrypoint (id);
+CREATE INDEX index_entrypoint_arg_entrypoint_id ON entrypoint_arg (entrypoint_id);
+
+
+CREATE TABLE entrypoint_commands (
+    devfile_entrypoint_id    BIGINT,
+    commands                 VARCHAR(255) NOT NULL
```

Comment:
It is commands or just a command?

Reply:
I wonder how JPA keeps the original order of command parts since we do not have the corresponding field (I mean order index) in database. Maybe records are ordered by created timestamp.
I checked and it really works, but would be nice to deeply understand why
```java  @Test(dependsOnMethods = "shouldGetWorkspaceById")
  public void shouldKeepOriginalCommandPartsOrder() throws Exception {
    final WorkspaceImpl workspace =
        createWorkspaceFromDevfile("new-workspace", accounts[1], "new-name");

    workspace.getDevfile().getComponents().get(0).setCommand(ImmutableList.of("sh", "-c", "echo hello"));
    workspaceDao.create(workspace);

    WorkspaceImpl actual = workspaceDao.get(workspace.getId());
    assertEquals(
        actual, new WorkspaceImpl(workspace, workspace.getAccount()));
    List<String> command = actual.getDevfile().getComponents().get(0).getCommand();
    assertEquals(command.get(0), "sh");
    assertEquals(command.get(1), "-c");
    assertEquals(command.get(2), "echo hello");
  }
```